IdentityCore/src/requests/broker/MSIDGCDStarvationDetector.m

```diff
@@ -0,0 +1,127 @@
+//
+// Copyright (c) Microsoft Corporation.
+// All rights reserved.
+//
+// This code is licensed under the MIT License.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files(the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and / or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions :
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+#import "MSIDGCDStarvationDetector.h"
+#import "MSIDLogger+Internal.h"
+
+static NSTimeInterval starvationCheckTimeout = 0.01; //10 ms - timeout for detecting if GCD thread pool is starved
+static NSTimeInterval starvationCheckInterval = 0.1; //100 ms - interval between starvation checks
+static NSTimeInterval maxMonitoringDuration = 15.0; //15 seconds - maximum monitoring duration to prevent indefinite running
+
+@interface MSIDGCDStarvationDetector()
+
+@property (nonatomic) BOOL running;
+@property (nonatomic) BOOL shouldStop;
```

Comment:
Seems these properties will be accessed by multiple threads. Read and Write at teh same time will cause crash. Shall we make them atomic?